File: src/test/java/io/github/bot/terminal/application/domains/notificarion_api/rest/NotificationApiRestServiceTest.java
package io.github.bot.terminal.application.domains.notificarion_api.rest;

import io.github.bot.terminal.application.domains.notificarion_api.entity.*;
import io.github.bot.terminal.application.domains.notificarion_api.factory.NotificationApiFactory;
import io.github.bot.terminal.application.domains.notificarion_api.repository.NotificationApiRepository;
import io.github.bot.terminal.application.domains.notificarion_api.rest.dto.AbstractNotificationApiDTO;
import io.github.bot.terminal.application.domains.notificarion_api.rest.dto.NotificationApiTypeDTO;
import io.github.bot.terminal.application.domains.notificarion_api.rest.dto.SlackNotificationApiDTO;
import io.github.bot.terminal.application.domains.notificarion_api.rest.dto.TelegramNotificationApiDTO;
import io.github.bot.terminal.application.domains.notificarion_api.rest.requests.SlackNotificationApiRequest;
import io.github.bot.terminal.application.domains.notificarion_api.rest.requests.TelegramNotificationApiRequest;
import io.github.bot.terminal.application.domains.notificarion_api.values.NotificationApiState;
import io.github.bot.terminal.application.domains.notificarion_api.values.NotificationApiType;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;


@ExtendWith(MockitoExtension.class)
class NotificationApiRestServiceTest {

    @Mock
    private NotificationApiFactory notificationApiFactory;
    @Mock
    private NotificationApiRepository repository;

    private NotificationApiRestConverter converter = spy(new NotificationApiRestConverter());
    private NotificationApiRestService restService;

    @Captor
    ArgumentCaptor<SlackNotificationApiDetails> slackDetailsCaptor;
    @Captor
    ArgumentCaptor<TelegramNotificationApiDetails> telegramDetailsCaptor;

    private String id = "id";
    private String label = "label";
    private NotificationApiState state = NotificationApiState.ENABLED;
    private String chanel = "chanel-name";
    private String botFatherName = "botFatherName-name";
    private String token = "token";

    private String id2 = "id2";
    private String label2 = "label";
    private NotificationApiState state2 = NotificationApiState.DISABLED;
    private String chanel2 = "chanel-name2";
    private String botFatherName2 = "botFatherName-name2";
    private String token2 = "token2";


    @BeforeEach
    public void inti() {
        restService = new NotificationApiRestService(notificationApiFactory, converter);
    }


    @Test
    public void addUnknownNotificationApi() {
        SlackNotificationApiRequest request = new SlackNotificationApiRequest();
        request.setLabel(label);
        request.setType("SomeType");
        request.setToken(token);
        request.setChanel(chanel);
        request.setState(state.name());

        assertThrows(IllegalArgumentException.class, () ->
                restService.add(request)
        );

    }

    @Test
    public void mapFromUnknownNotificationApiType() {

        SlackNotificationApiDetails details = new SlackNotificationApiDetails();
        details.setId(id);
        details.setLabel(label);
        details.setType(null);
        details.setState(state);
        details.setChanel(chanel);
        details.setToken(token);

        NotificationApi api = new SlackNotificationApi(details, repository);

        when(notificationApiFactory.createNew(any()))
                .thenReturn(api);

        SlackNotificationApiRequest request = new SlackNotificationApiRequest();
        request.setLabel(label);
        request.setToken(token);
        request.setChanel(chanel);
        request.setState(state.name());

        assertThrows(IllegalArgumentException.class, () ->
                restService.add(request)
        );

    }

    @Test
    public void addNewSlackNotificationApi() {

        SlackNotificationApiDetails details = new SlackNotificationApiDetails();
        details.setId(id);
        details.setLabel(label);
        details.setType(NotificationApiType.SLACK_BOT);
        details.setState(state);
        details.setChanel(chanel);
        details.setToken(token);

        NotificationApi api = new SlackNotificationApi(details, repository);

        when(notificationApiFactory.createNew(any()))
                .thenReturn(api);

        SlackNotificationApiRequest request = new SlackNotificationApiRequest();
        request.setLabel(label);
        request.setToken(token);
        request.setChanel(chanel);
        request.setState(state.name());

        SlackNotificationApiDTO apiDTO = (SlackNotificationApiDTO) restService.add(request);
        assertEquals(id, apiDTO.getId());
        assertEquals(label, apiDTO.getLabel());
        assertEquals(state.name(), apiDTO.getState());
        assertEquals(token, apiDTO.getToken());
        assertEquals(chanel, apiDTO.getChanel());
        assertEquals(NotificationApiType.SLACK_BOT.name(), apiDTO.getType());

        verify(notificationApiFactory, times(1)).createNew(slackDetailsCaptor.capture());
        SlackNotificationApiDetails convertedDetails = slackDetailsCaptor.getValue();
        assertEquals(label, convertedDetails.getLabel());
        assertEquals(state, convertedDetails.getState());
        assertEquals(token, convertedDetails.getToken());
        assertEquals(chanel, convertedDetails.getChanel());
        assertEquals(NotificationApiType.SLACK_BOT, convertedDetails.getType());
    }


    @Test
    public void editSlackNotificationApi() {

        SlackNotificationApiDetails details = new SlackNotificationApiDetails();
        details.setId(id);
        details.setLabel(label2);
        details.setType(NotificationApiType.SLACK_BOT);
        details.setState(state2);
        details.setChanel(chanel2);
        details.setToken(token2);

        NotificationApi api = new SlackNotificationApi(details, repository);

        when(notificationApiFactory.merge(anyString(), any()))
                .thenReturn(api);

        SlackNotificationApiRequest request = new SlackNotificationApiRequest();
        request.setLabel(label2);
        request.setToken(token2);
        request.setChanel(chanel2);
        request.setState(state2.name());

        SlackNotificationApiDTO apiDTO = (SlackNotificationApiDTO) restService.edit(id, request);
        assertEquals(id, apiDTO.getId());
        assertEquals(label2, apiDTO.getLabel());
        assertEquals(state2.name(), apiDTO.getState());
        assertEquals(token2, apiDTO.getToken());
        assertEquals(chanel2, apiDTO.getChanel());
        assertEquals(NotificationApiType.SLACK_BOT.name(), apiDTO.getType());

        verify(notificationApiFactory, times(1)).merge(eq(id), slackDetailsCaptor.capture());
        SlackNotificationApiDetails convertedDetails = slackDetailsCaptor.getValue();
        assertEquals(label2, convertedDetails.getLabel());
        assertEquals(state2, convertedDetails.getState());
        assertEquals(token2, convertedDetails.getToken());
        assertEquals(chanel2, convertedDetails.getChanel());
        assertEquals(NotificationApiType.SLACK_BOT, convertedDetails.getType());
    }

    @Test
    public void getSlackNotificationApi() {

        SlackNotificationApiDetails details = new SlackNotificationApiDetails();
        details.setId(id);
        details.setLabel(label);
        details.setType(NotificationApiType.SLACK_BOT);
        details.setState(state);
        details.setChanel(chanel);
        details.setToken(token);

        NotificationApi api = new SlackNotificationApi(details, repository);

        when(notificationApiFactory.byId(eq(id)))
                .thenReturn(api);

        SlackNotificationApiDTO apiDTO = (SlackNotificationApiDTO) restService.get(id);
        assertEquals(id, apiDTO.getId());
        assertEquals(label, apiDTO.getLabel());
        assertEquals(state.name(), apiDTO.getState());
        assertEquals(token, apiDTO.getToken());
        assertEquals(chanel, apiDTO.getChanel());
        assertEquals(NotificationApiType.SLACK_BOT.name(), apiDTO.getType());

    }


    @Test
    public void addNewTelegramNotificationApi() {

        TelegramNotificationApiDetails details = new TelegramNotificationApiDetails();
        details.setId(id);
        details.setLabel(label);
        details.setType(NotificationApiType.TELEGRAM_BOT);
        details.setState(state);
        details.setBotFatherName(botFatherName);
        details.setToken(token);

        NotificationApi api = new TelegramNotificationApi(details, repository);

        when(notificationApiFactory.createNew(any()))
                .thenReturn(api);

        TelegramNotificationApiRequest request = new TelegramNotificationApiRequest();
        request.setLabel(label);
        request.setToken(token);
        request.setBotFatherName(botFatherName);
        request.setState(state.name());

        TelegramNotificationApiDTO apiDTO = (TelegramNotificationApiDTO) restService.add(request);
        assertEquals(id, apiDTO.getId());
        assertEquals(label, apiDTO.getLabel());
        assertEquals(state.name(), apiDTO.getState());
        assertEquals(token, apiDTO.getToken());
        assertEquals(botFatherName, apiDTO.getBotFatherName());
        assertEquals(NotificationApiType.TELEGRAM_BOT.name(), apiDTO.getType());

        verify(notificationApiFactory, times(1)).createNew(telegramDetailsCaptor.capture());
        TelegramNotificationApiDetails convertedDetails = telegramDetailsCaptor.getValue();
        assertEquals(label, convertedDetails.getLabel());
        assertEquals(state, convertedDetails.getState());
        assertEquals(token, convertedDetails.getToken());
        assertEquals(botFatherName, convertedDetails.getBotFatherName());
        assertEquals(NotificationApiType.TELEGRAM_BOT, convertedDetails.getType());
    }

    @Test
    public void editTelegramNotificationApi() {

        TelegramNotificationApiDetails details = new TelegramNotificationApiDetails();
        details.setId(id);
        details.setLabel(label2);
        details.setType(NotificationApiType.TELEGRAM_BOT);
        details.setState(state2);
        details.setBotFatherName(botFatherName2);
        details.setToken(token2);

        NotificationApi api = new TelegramNotificationApi(details, repository);

        when(notificationApiFactory.merge(anyString(), any()))
                .thenReturn(api);

        TelegramNotificationApiRequest request = new TelegramNotificationApiRequest();
        request.setLabel(label2);
        request.setToken(token2);
        request.setBotFatherName(botFatherName2);
        request.setState(state2.name());

        TelegramNotificationApiDTO apiDTO = (TelegramNotificationApiDTO) restService.edit(id, request);
        assertEquals(id, apiDTO.getId());
        assertEquals(label2, apiDTO.getLabel());
        assertEquals(state2.name(), apiDTO.getState());
        assertEquals(token2, apiDTO.getToken());
        assertEquals(botFatherName2, apiDTO.getBotFatherName());
        assertEquals(NotificationApiType.TELEGRAM_BOT.name(), apiDTO.getType());

        verify(notificationApiFactory, times(1)).merge(eq(id), telegramDetailsCaptor.capture());
        TelegramNotificationApiDetails convertedDetails = telegramDetailsCaptor.getValue();
        assertEquals(label2, convertedDetails.getLabel());
        assertEquals(state2, convertedDetails.getState());
        assertEquals(token2, convertedDetails.getToken());
        assertEquals(botFatherName2, convertedDetails.getBotFatherName());
        assertEquals(NotificationApiType.TELEGRAM_BOT, convertedDetails.getType());
    }


    @Test
    public void getTelegramNotificationApi() {

        TelegramNotificationApiDetails details = new TelegramNotificationApiDetails();
        details.setId(id);
        details.setLabel(label);
        details.setType(NotificationApiType.TELEGRAM_BOT);
        details.setState(state);
        details.setBotFatherName(botFatherName);
        details.setToken(token);

        NotificationApi api = new TelegramNotificationApi(details, repository);

        when(notificationApiFactory.byId(eq(id)))
                .thenReturn(api);

        TelegramNotificationApiDTO apiDTO = (TelegramNotificationApiDTO) restService.get(id);
        assertEquals(id, apiDTO.getId());
        assertEquals(label, apiDTO.getLabel());
        assertEquals(state.name(), apiDTO.getState());
        assertEquals(token, apiDTO.getToken());
        assertEquals(botFatherName, apiDTO.getBotFatherName());
        assertEquals(NotificationApiType.TELEGRAM_BOT.name(), apiDTO.getType());
    }


    @Test
    public void deleteNotificationApi() {

        NotificationApi api = Mockito.mock(NotificationApi.class);

        when(notificationApiFactory.byId(eq(id)))
                .thenReturn(api);

        restService.delete(id);

        verify(api, timeout(1)).delete();
    }


    @Test
    public void getAllNotificationApi() {

        SlackNotificationApiDetails details = new SlackNotificationApiDetails();
        details.setId(id);
        details.setLabel(label);
        details.setType(NotificationApiType.SLACK_BOT);
        details.setState(state);
        details.setChanel(chanel);
        details.setToken(token);

        NotificationApi api1 = new SlackNotificationApi(details, repository);

        TelegramNotificationApiDetails details2 = new TelegramNotificationApiDetails();
        details2.setId(id2);
        details2.setLabel(label2);
        details2.setType(NotificationApiType.TELEGRAM_BOT);
        details2.setState(state2);
        details2.setBotFatherName(botFatherName2);
        details2.setToken(token2);

        NotificationApi api2 = new TelegramNotificationApi(details2, repository);


        List<NotificationApi<?>> c = new ArrayList();
        c.add(api1);
        c.add(api2);

        when(notificationApiFactory.getAll())
                .thenReturn(c);

        List<AbstractNotificationApiDTO> list = restService.list();

        SlackNotificationApiDTO apiDTO = (SlackNotificationApiDTO)  list.get(0);
        assertEquals(id, apiDTO.getId());
        assertEquals(label, apiDTO.getLabel());
        assertEquals(state.name(), apiDTO.getState());
        assertEquals(token, apiDTO.getToken());
        assertEquals(chanel, apiDTO.getChanel());
        assertEquals(NotificationApiType.SLACK_BOT.name(), apiDTO.getType());

        TelegramNotificationApiDTO apiDTO2 = (TelegramNotificationApiDTO) list.get(1);
        assertEquals(id2, apiDTO2.getId());
        assertEquals(label2, apiDTO2.getLabel());
        assertEquals(state2.name(), apiDTO2.getState());
        assertEquals(token2, apiDTO2.getToken());
        assertEquals(botFatherName2, apiDTO2.getBotFatherName());
        assertEquals(NotificationApiType.TELEGRAM_BOT.name(), apiDTO2.getType());

    }


    @Test
    public void getTypes() {
        List<NotificationApiTypeDTO> types = restService.types();
        assertEquals(2, types.size());
        assertEquals(NotificationApiType.TELEGRAM_BOT.name(), types.get(0).getType());
        assertEquals(NotificationApiType.TELEGRAM_BOT.getLabel(), types.get(0).getLabel());
        assertEquals(NotificationApiType.SLACK_BOT.name(), types.get(1).getType());
        assertEquals(NotificationApiType.SLACK_BOT.getLabel(), types.get(1).getLabel());
    }
}

File: src/test/kotlin/io/github/bot/terminal/application/domains/notificarion_api/rest/NotificationApiRestServiceTest.kt
package io.github.bot.terminal.application.domains.notificarion_api.rest

import com.nhaarman.mockitokotlin2.*
import io.github.bot.terminal.application.domains.notificarion_api.NotificationApiTestData
import io.github.bot.terminal.application.domains.notificarion_api.NotificationsApiDataSet
import io.github.bot.terminal.application.domains.notificarion_api.entity.NotificationApiDetails
import io.github.bot.terminal.application.domains.notificarion_api.entity.SlackNotificationApiDetails
import io.github.bot.terminal.application.domains.notificarion_api.entity.TelegramNotificationApiDetails
import io.github.bot.terminal.application.domains.notificarion_api.factory.NotificationApiFactory
import io.github.bot.terminal.application.domains.notificarion_api.rest.dto.NotificationApiDTO
import io.github.bot.terminal.application.domains.notificarion_api.rest.dto.SlackNotificationApiDTO
import io.github.bot.terminal.application.domains.notificarion_api.rest.dto.TelegramNotificationApiDTO
import io.github.bot.terminal.application.domains.notificarion_api.values.NotificationApiType
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.ArgumentCaptor
import org.mockito.Captor
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
internal class NotificationApiRestServiceTest {

    private val notificationApiFactory: NotificationApiFactory = mock()

    @Captor
    private lateinit var detailsCaptor: ArgumentCaptor<NotificationApiDetails>

    private val converter = Mockito.spy(NotificationApiRestConverter())

    private lateinit var restService: NotificationApiRestService

    @BeforeEach
    fun inti() {
        reset(notificationApiFactory)
        restService = NotificationApiRestService(notificationApiFactory, converter)
    }

    @Test
    fun addUnknownNotificationApi() {
        Assertions.assertThrows(IllegalArgumentException::class.java) {
            restService.add(NotificationsApiDataSet.Wrong.WRONG_1.request)
        }
    }

    @Test
    fun `add new slack notificationApi`() {
        addNewNotificationApi(NotificationsApiDataSet.Slack.SLACK_1,
                {
                    val dto = it as SlackNotificationApiDTO
                    val data = NotificationsApiDataSet.Slack.SLACK_1
                    Assertions.assertEquals(data.id, dto.id)
                    Assertions.assertEquals(data.label, dto.label)
                    Assertions.assertEquals(data.enabled, dto.enabled)
                    Assertions.assertEquals(data.token, dto.token)
                    Assertions.assertEquals(data.chanel, dto.chanel)
                },
                {
                    val details = it as SlackNotificationApiDetails
                    val data = NotificationsApiDataSet.Slack.SLACK_1
                    Assertions.assertEquals(data.label, details.label)
                    Assertions.assertEquals(data.enabled, details.enabled)
                    Assertions.assertEquals(data.token, details.token)
                    Assertions.assertEquals(data.chanel, details.chanel)
                }
        )
    }

    @Test
    fun `add new telegram notificationApi`() {
        addNewNotificationApi(NotificationsApiDataSet.Telegram.TELEGRAM_1,
                {
                    val dto = it as TelegramNotificationApiDTO
                    val data = NotificationsApiDataSet.Telegram.TELEGRAM_1
                    Assertions.assertEquals(dto.id, data.id)
                    Assertions.assertEquals(dto.label, data.label)
                    Assertions.assertEquals(dto.enabled, data.enabled)
                    Assertions.assertEquals(dto.token, data.token)
                    Assertions.assertEquals(dto.botFatherName, data.botFatherName)
                },
                {
                    val details = it as TelegramNotificationApiDetails
                    val data = NotificationsApiDataSet.Telegram.TELEGRAM_1
                    Assertions.assertEquals(data.label, details.label)
                    Assertions.assertEquals(data.enabled, details.enabled)
                    Assertions.assertEquals(data.token, details.token)
                    Assertions.assertEquals(data.botFatherName, details.botFatherName)
                }
        )
    }

    private fun addNewNotificationApi(data: NotificationApiTestData,
                                      compareDto: (NotificationApiDTO) -> Unit,
                                      compareDetails: (NotificationApiDetails) -> Unit) {
        whenever(notificationApiFactory.createNew(any()))
                .thenReturn(data.api())

        val dto = restService.add(data.request())
        compareDto(dto)

        verify(notificationApiFactory, times(1)).createNew(capture(detailsCaptor))
        compareDetails(detailsCaptor.value)
    }

    @Test
    fun `edit slack notificationApi`() {
        editNotificationApi(NotificationsApiDataSet.Slack.SLACK_1,
                NotificationsApiDataSet.Slack.SLACK_1_UPDATE,
                {
                    val dto = it as SlackNotificationApiDTO
                    val update = NotificationsApiDataSet.Slack.SLACK_1_UPDATE
                    Assertions.assertEquals(update.id, dto.id)
                    Assertions.assertEquals(update.label, dto.label)
                    Assertions.assertEquals(update.enabled, dto.enabled)
                    Assertions.assertEquals(update.token, dto.token)
                    Assertions.assertEquals(update.chanel, dto.chanel)
                },
                {
                    val details = it as SlackNotificationApiDetails
                    val updateDetails = NotificationsApiDataSet.Slack.SLACK_1_UPDATE.details()
                    Assertions.assertEquals(updateDetails.label, details.label)
                    Assertions.assertEquals(updateDetails.enabled, details.enabled)
                    Assertions.assertEquals(updateDetails.token, details.token)
                    Assertions.assertEquals(updateDetails.chanel, details.chanel)
                }
        )
    }

    @Test
    fun `edit telegram notificationApi`() {
        editNotificationApi(NotificationsApiDataSet.Telegram.TELEGRAM_1,
                NotificationsApiDataSet.Telegram.TELEGRAM_1_UPDATE,
                {
                    val dto = it as TelegramNotificationApiDTO
                    val update = NotificationsApiDataSet.Telegram.TELEGRAM_1_UPDATE
                    Assertions.assertEquals(dto.id, update.id)
                    Assertions.assertEquals(dto.label, update.label)
                    Assertions.assertEquals(dto.enabled, update.enabled)
                    Assertions.assertEquals(dto.token, update.token)
                    Assertions.assertEquals(dto.botFatherName, update.botFatherName)
                }, {
            val details = it as TelegramNotificationApiDetails
            val updateDetails = NotificationsApiDataSet.Telegram.TELEGRAM_1_UPDATE.details()
            Assertions.assertEquals(updateDetails.label, details.label)
            Assertions.assertEquals(updateDetails.enabled, details.enabled)
            Assertions.assertEquals(updateDetails.token, details.token)
            Assertions.assertEquals(updateDetails.botFatherName, details.botFatherName)
        })
    }

    private fun editNotificationApi(data: NotificationApiTestData, update: NotificationApiTestData,
                                    compareDto: (NotificationApiDTO) -> Unit,
                                    compareDetails: (NotificationApiDetails) -> Unit) {
        whenever(notificationApiFactory.update(any(), any()))
                .thenReturn(update.api())
        val dto = restService.edit(data.id(), update.request())
        compareDto(dto)
        verify(notificationApiFactory, times(1)).update(eq(data.id()), capture(detailsCaptor))
        compareDetails(detailsCaptor.value)
    }

    @Test
    fun `get notificationApi types`() {
        val types = restService.types()
        Assertions.assertEquals(2, types.size)
        Assertions.assertEquals(NotificationApiType.TELEGRAM_BOT.name, types[0].type)
        Assertions.assertEquals(NotificationApiType.TELEGRAM_BOT.label, types[0].label)
        Assertions.assertEquals(NotificationApiType.SLACK_BOT.name, types[1].type)
        Assertions.assertEquals(NotificationApiType.SLACK_BOT.label, types[1].label)
    }

    @Test
    fun `delete notificationApi`() {
        restService.delete(NotificationsApiDataSet.Slack.SLACK_1.id)
        verify(notificationApiFactory, Mockito.timeout(1)).delete(eq(NotificationsApiDataSet.Slack.SLACK_1.id))
    }

    @Test
    fun `get all notificationApi`() {
        val apiList = listOf(
                NotificationsApiDataSet.Slack.SLACK_1.api(),
                NotificationsApiDataSet.Telegram.TELEGRAM_1.api()
        )
        whenever(notificationApiFactory.all()).thenReturn(apiList)

        val dtoList = listOf(
                NotificationsApiDataSet.Slack.SLACK_1.dto(),
                NotificationsApiDataSet.Telegram.TELEGRAM_1.dto()
        )
        val resultList = restService.list()
        Assertions.assertEquals(dtoList, resultList)
    }

    @Test
    fun `get slack notificationApi`() {
        getNotificationApi(NotificationsApiDataSet.Slack.SLACK_1)
    }

    @Test
    fun `get telegram notificationApi`() {
        getNotificationApi(NotificationsApiDataSet.Telegram.TELEGRAM_1)
    }

    fun getNotificationApi(data: NotificationApiTestData) {
        whenever(notificationApiFactory.byId(eq(data.id()))).thenReturn(data.api())
        val result = restService.get(data.id())
        Assertions.assertEquals(data.dto(), result)
    }

}
